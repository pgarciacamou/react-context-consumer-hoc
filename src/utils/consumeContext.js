import React from 'react'

/**
 * Wrap the inner consumer with the consumer from ContextAPI
 *
 * @param {Element} InnerConsumer
 * @param {Context} ContextAPI
 * @returns {Element} outer consumer
 */
export function consumeContext(InnerConsumer, ContextAPI) {
  // Ignore falsy values to prevent unhandled errors.
  if (!ContextAPI) {
    return InnerConsumer
  }

  return ({ context: parentContext = {}, render }) => (
    <ContextAPI.Consumer>
      {context => (
        <InnerConsumer
          render={render}
          context={{
            ...parentContext,
            ...context
          }}
        />
      )}
    </ContextAPI.Consumer>
  )
}
