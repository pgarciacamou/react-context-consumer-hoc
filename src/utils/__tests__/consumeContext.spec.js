import React from 'react'
import renderer from 'react-test-renderer'
import { consumeContext } from '../consumeContext'

describe('createContextConsumerChain', () => {
  it('matches snapshot', () => {
    const previousContext = { b: 2 }
    const childContextA = { a: 1 }
    const ContextA = React.createContext(childContextA)

    const Consumer = consumeContext(
      function InnerConsumer({ context, render, ...thisShouldBeEmpty }) {
        expect(thisShouldBeEmpty).toEqual({})
        return (<div {...context} render={render} {...thisShouldBeEmpty} />)
      },
      ContextA
    )

    const tree = renderer.create(
      <ContextA.Provider value={childContextA}>
        <Consumer context={previousContext} render={() => {}} />
      </ContextA.Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders inner consumer with context and passes render function', () => {
    const spy = jest.fn((context) => null)
    const previousContext = { b: 2 }
    const childContextA = { a: 1 }
    const consumedContext = { ...previousContext, ...childContextA }
    const ContextA = React.createContext(childContextA)

    const Consumer = consumeContext(
      function InnerConsumer({ context, render }) {
        expect(context).toEqual(consumedContext) // passes and adds context
        expect(render).toEqual(spy) // passes render
        return render(context)
      },
      ContextA
    )

    renderer.create(
      <ContextA.Provider value={childContextA}>
        <Consumer context={previousContext} render={spy} />
      </ContextA.Provider>
    )

    expect(spy).toHaveBeenCalledTimes(1) // must render the InnerConsumer
    expect(spy).toHaveBeenCalledWith(consumedContext) // render fn executed correctly
  })

  it('it will not override context but add it up', () => {
    const previousContext = { b: 2 }
    const childContextA = { a: 1 }
    const ContextA = React.createContext(childContextA)

    const InnerConsumer = ({ context }) => {
      expect(context).not.toEqual(previousContext) // did not override context prop
      expect(context).not.toEqual(childContextA) // did not override context prop
      return null
    }
    const Consumer = consumeContext(InnerConsumer, ContextA)

    renderer.create(
      <ContextA.Provider value={childContextA}>
        <Consumer context={previousContext} />
      </ContextA.Provider>
    )
  })
})
