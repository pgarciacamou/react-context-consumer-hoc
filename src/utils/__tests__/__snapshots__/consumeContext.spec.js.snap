// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContextConsumerChain matches snapshot 1`] = `
<div
  a={1}
  b={2}
  render={[Function]}
/>
`;
